File: 707-design-linked-list/707-design-linked-list.py
class Node:
    def __init__(self, val, next=None):
        self.val = val
        self.next = next
        
    def print(self):
        if not self.next:
            print(self.val)
        else:
            print(self.val, end=" ")
            self.next.print()

class MyLinkedList:

    def __init__(self):
        self.head = Node(None)
        self.tail = self.head
        self.length = 0

    def get(self, index: int) -> int:
        if index >= self.length or index < 0:
            return -1
        
        curr = self.head
        for _ in range(index + 1):
            curr = curr.next
        
        return curr.val

    def addAtHead(self, val: int) -> None:
        self.head.next = Node(val, self.head.next)
        self.length += 1
        
        if self.tail == self.head:
            self.tail = self.head.next

    def addAtTail(self, val: int) -> None:
        self.tail.next = Node(val)
        self.tail = self.tail.next
        self.length += 1
        
    def addAtIndex(self, index: int, val: int) -> None:
        if index > self.length or index < 0:
            return 
        
        curr = self.head
        for _ in range(index):
            curr = curr.next
            
        curr.next = Node(val, curr.next)
        self.length += 1
        
        if self.tail == curr:
            self.tail = curr.next

    def deleteAtIndex(self, index: int) -> None:
        if index >= self.length or index < 0:
            return 
        
        if self.length == 0:
            return
        
        curr = self.head
        for _ in range(index):
            curr = curr.next
        
        if curr.next == self.tail:
            curr.next = None
            self.tail = curr
        else:
            curr.next = curr.next.next
        
        self.length -= 1

# Your MyLinkedList object will be instantiated and called as such:
# obj = MyLinkedList()
# param_1 = obj.get(index)
# obj.addAtHead(val)
# obj.addAtTail(val)
# obj.addAtIndex(index,val)
# obj.deleteAtIndex(index)

File: 707-design-linked-list/707-design-linked-list.cpp
struct Node {
    int val;
    Node* next;
    Node* prev;
    
    Node(int val) {
        this->val = val;
        this->next = nullptr;
    }
    
    Node(int val, Node* prev, Node* next) {
        this->val = val;
        this->prev = prev;
        this->next = next;
    }
};

class MyLinkedList {
private:
    int length;
    Node* head;
    Node* tail;
public:
    MyLinkedList() {
        this->length = 0;
        this->head = new Node(0);
        this->tail = new Node(0);
        this->head->next = this->tail;
        this->tail->prev = this->head;
    }
    
    int get(int index) {
        if (index >= this->length || index < 0) {
            return -1;
        }
        
        Node* curr = this->head;
        for (int i = 0; i <= index; i++) {
            curr = curr->next;
        }
        return curr->val;
    }
    
    void addAtHead(int val) {
        addAtIndex(0, val);
    }
    
    void addAtTail(int val) {
        addAtIndex(this->length, val);
    }
    
    void addAtIndex(int index, int val) {
        if (index > this->length || index < 0) {
            return;
        }
        
        Node* curr = this->head;
        for (int i = 0; i < index; i++) {
            curr = curr->next;
        }
        
        Node* newNode = new Node(val, curr, curr->next);
        curr->next = newNode;
        newNode->next->prev = newNode;

        this->length++;
    }
    
    void deleteAtIndex(int index) {
        if (index >= this->length || index < 0) {
            return;
        }
        if (this->length == 0) {
            return;
        }
        
        Node* curr = this->head;
        for (int i = 0; i <= index; i++) {
            curr = curr->next;
        }
        
        Node* toRemove = curr;
        toRemove->prev->next = toRemove->next;
        toRemove->next->prev = toRemove->prev;

        // delete toRemove;
        this->length--;
    }
};

/**
 * Your MyLinkedList object will be instantiated and called as such:
 * MyLinkedList* obj = new MyLinkedList();
 * int param_1 = obj->get(index);
 * obj->addAtHead(val);
 * obj->addAtTail(val);
 * obj->addAtIndex(index,val);
 * obj->deleteAtIndex(index);
 */